### ISSUE TITLE
ProgressListenerCallbackExecutor prevents back-pressure, and may cause deadlock

### ISSUE THREAD

**Issue by joshng** [aws/aws-sdk-java#145]
Since v1.6, all ProgressListener callbacks are funneled through a global, single-threaded, unbounded ExecutorService. This implies that if any ProgressListener invocation blocks, then all subsequent callbacks -- process-wide -- will be blocked as well, which can cause subtle deadlocks. Moreover, if downstream bottlenecks do arise and cause callbacks to block, the unbounded queue employed by this ExecutorService may continue to fill with pending callbacks until memory overflow occurs.

The shared global ExecutorService can create a hidden cycle in what might otherwise be a DAG of stages in an asynchronous processing pipeline. This effectively prevents a prudent designer from allowing threads to block for back-pressure.

In short, with the current implementation, blocking any ProgressListener callback is a dangerous practice in a high-volume application, but avoiding blocking generally requires more unbounded queues -- also dangerous -- or load-shedding (not always an option).

Ideally, the application should be able to supply an Executor along with each ProgressListener to customize the threading of the asynchronous callbacks.

I've prototyped a quick backwards-compatible workaround in my [progress-listener-with-executor](https://github.com/joshng/aws-sdk-java/tree/progress-listener-with-executor) branch, but I'll note that a better solution would surface the Executor specification in a more visible manner via the API.

**Comment by hanshuo-aws** [comment]
Sorry for the silence on this.

In v1.6, we changed all the ProgressListener callbacks to be asynchronously executed after we got customer reports showing that it could lead to deadlock in some edge case when you supply a blocking listener callback to the TransferManager (where ProgressListener is mostly used).

However, because of the exact same reasons you mentioned here, we later found this async approach caused even bigger problems compared to the specific issue we were trying to fix. So as a trade-off, we added the [DeliveryMode](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/event/DeliveryMode.java) interface to the SDK, which allows you to explicitly control the execution strategy for a particular listener callback by implementing the <code>isSyncCallSafe</code> method.

I think this mechanism has solved the fundamental issue here, i.e.,  whether a callback should be executed sync or async should be configurable according to the use case and that the SDK should not enforce a global default behavior. I hope this answers your question and concern, and please let use know if you have any further comment.
